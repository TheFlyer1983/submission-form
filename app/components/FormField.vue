<script lang="ts" setup>
const modelValue = defineModel<string | boolean>();
const otherValue = defineModel<string>('otherValue');

const props = defineProps<{
  label: string;
  type: string;
  required: boolean;
  error?: string;
}>();

const { data } = await useLazyFetch('/api/services');
</script>

<template>
  <div>
    <div
      v-if="
        ['text', 'password', 'date'].includes(props.type) &&
        typeof modelValue === 'string'
      "
    >
      <label
        >{{ label }}: <span v-if="required" class="text-red-600">*</span></label
      >
      <UiInput v-model="modelValue" :type />
    </div>
    <div
      v-else-if="
        ['select'].includes(props.type) && typeof modelValue === 'string'
      "
      class="space-y-3"
    >
      <div>
        <label> {{ label }}: <span v-if="required">*</span> </label>
        <UiSelect v-model="modelValue">
          <UiSelectTrigger>
            <UiSelectValue placeholder="Please select a service" />
          </UiSelectTrigger>
          <UiSelectContent>
            <UiSelectGroup>
              <UiSelectItem
                v-for="item in data"
                :key="item.value"
                :value="item.value"
              >
                {{ item.label }}
              </UiSelectItem>
            </UiSelectGroup>
          </UiSelectContent>
        </UiSelect>
      </div>
      <div v-if="modelValue === 'other'">
        <label>Other:</label>
        <UiInput v-model="otherValue" />
      </div>
    </div>
    <div
      v-else-if="
        ['checkbox'].includes(props.type) && typeof modelValue === 'boolean'
      "
    >
      <UiCheckbox v-model:checked="modelValue" />
      <label class="ml-2">{{ label }}</label>
    </div>
    <div class= text-red-500">{{ error }}</div>
  </div>
</template>
