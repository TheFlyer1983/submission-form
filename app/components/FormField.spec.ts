import { mountSuspended, registerEndpoint } from '@nuxt/test-utils/runtime';

import FormField from './FormField.vue';
import UiInput from '@/components/ui/input/Input.vue';
import UiSelect from '@/components/ui/select/Select.vue';
import UiCheckbox from '@/components/ui/checkbox/Checkbox.vue';

registerEndpoint('/api/services?locale=en', () => [
  {
    label: 'Web Development',
    value: 'web_development'
  },
  {
    label: 'Mobile Development',
    value: 'mobile_development'
  },
  {
    label: 'SEO Services',
    value: 'seo_services'
  },
  {
    label: 'Other',
    value: 'other'
  }
]);

describe('FormField', () => {
  describe('when the field type is text', () => {
    it('should render correctly', async () => {
      const wrapper = await mountSuspended(FormField, {
        props: {
          modelValue: 'Paul',
          field: {
            label: 'Name',
            name: 'name',
            type: 'text',
            required: true,
            parent: null,
            visible: true,
            validationRules: {
              required: true,
              minLength: 2,
              alpha: true
            }
          },
          visible: true,
          locale: 'en'
        }
      });

      expect(wrapper.html()).toMatchSnapshot();
      expect(wrapper.findComponent(UiInput).exists()).toBe(true);
      expect(wrapper.findComponent(UiSelect).exists()).toBe(false);
      expect(wrapper.findComponent(UiCheckbox).exists()).toBe(false);
    });
  });

  describe('when the field type is date', () => {
    it('should render correctly', async () => {
      const wrapper = await mountSuspended(FormField, {
        props: {
          modelValue: 'Paul',
          field: {
            label: 'Date of Birth',
            name: 'date_of_birth',
            type: 'date',
            required: false,
            parent: null,
            visible: true,
            validationRules: {
              isBefore: true
            }
          },
          visible: true,
          locale: 'en'
        }
      });

      expect(wrapper.html()).toMatchSnapshot();
      expect(wrapper.findComponent(UiInput).exists()).toBe(true);
      expect(wrapper.findComponent(UiSelect).exists()).toBe(false);
      expect(wrapper.findComponent(UiCheckbox).exists()).toBe(false);
    });
  });

  describe('when the field type is select', () => {
    it('should render correctly', async () => {
      const wrapper = await mountSuspended(FormField, {
        props: {
          modelValue: 'Paul',
          field: {
            label: 'Services',
            name: 'services',
            type: 'select',
            required: false,
            visible: true
          },
          visible: true,
          locale: 'en'
        }
      });

      expect(wrapper.html()).toMatchSnapshot();
      expect(wrapper.findComponent(UiInput).exists()).toBe(false);
      expect(wrapper.findComponent(UiSelect).exists()).toBe(true);
      expect(wrapper.findComponent(UiCheckbox).exists()).toBe(false);
    });
  });

  describe('when the field type is checkbox', () => {
    it('should render correctly', async () => {
      const wrapper = await mountSuspended(FormField, {
        props: {
          modelValue: true,
          field: {
            label:
              'Please check this box to agree with our terms and conditions',
            name: 'terms_and_conditions',
            type: 'checkbox',
            required: true,
            parent: null,
            visible: true,
            validationRules: {
              checked: true
            }
          },
          visible: true,
          locale: 'en'
        }
      });

      expect(wrapper.html()).toMatchSnapshot();
      expect(wrapper.findComponent(UiInput).exists()).toBe(false);
      expect(wrapper.findComponent(UiSelect).exists()).toBe(false);
      expect(wrapper.findComponent(UiCheckbox).exists()).toBe(true);
    });
  });
});
