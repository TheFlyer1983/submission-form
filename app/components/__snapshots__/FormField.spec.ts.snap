// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FormField > when the field type is checkbox > should render correctly 1`] = `
"<div>
  <div><button class="peer h-4 w-4 shrink-0 rounded-sm border border-primary ring-offset-background focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 data-[state=checked]:bg-primary data-[state=checked]:text-primary-foreground" role="checkbox" type="button" aria-checked="true" aria-required="false" data-state="checked"><span data-state="checked" style="pointer-events: none;" class="flex h-full w-full items-center justify-center text-current"><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="lucide lucide-check-icon h-4 w-4"><path d="M20 6 9 17l-5-5"></path></svg></span></button>
    <!----><label class="ml-2">Please check this box to agree with our terms and conditions</label>
  </div>
  <div class="text-red-500"></div>
</div>"
`;

exports[`FormField > when the field type is date > should render correctly 1`] = `
"<div>
  <div><label>Date of Birth:
      <!--v-if-->
    </label><input class="flex h-10 w-full rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50" type="date"></div>
  <div class="text-red-500"></div>
</div>"
`;

exports[`FormField > when the field type is select > should render correctly 1`] = `
"<div>
  <div class="space-y-3">
    <div><label>Services:
        <!--v-if-->
      </label><button role="combobox" type="button" aria-controls="radix-vue-select-content-v-0-0-0" aria-expanded="false" aria-required="false" aria-autocomplete="none" dir="ltr" data-state="closed" class="flex h-10 w-full items-center justify-between rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background data-[placeholder]:text-muted-foreground focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 [&amp;>span]:truncate text-start"><span style="pointer-events: none;"></span><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="lucide lucide-chevron-down-icon w-4 h-4 opacity-50 shrink-0" aria-hidden="true">
          <path d="m6 9 6 6 6-6"></path>
        </svg></button>
      <!--teleport start-->
      <!--teleport end--><select default-value="Paul" style="position: absolute; border: 0px; width: 1px; display: inline-block; height: 1px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0, 0, 0, 0); white-space: nowrap; word-wrap: normal;" aria-hidden="true" tabindex="-1" value="Paul"></select>
    </div>
  </div>
  <div class="text-red-500"></div>
</div>"
`;

exports[`FormField > when the field type is text > should render correctly 1`] = `
"<div>
  <div><label>Name: <span class="text-red-600">*</span></label><input class="flex h-10 w-full rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50" type="text"></div>
  <div class="text-red-500"></div>
</div>"
`;
